File: src/rv/models/manifest.py
"""Pydantic schemas for Manifest, Assets, Secrets, Packages, and Profiles.

Enforces strict v2 validation and domain-specific rules.
"""

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field, field_validator, model_validator

# Supported manifest schema versions. Raise UnsupportedSchemaVersionError for anything else.
_SUPPORTED_SCHEMA_VERSIONS: frozenset[int] = frozenset({1, 2})


class UnsupportedSchemaVersionError(ValueError):
    """Raised when a manifest declares a schema version that this rv release does not support.

    Prevents silent data corruption when a future schema adds conflicting or renamed fields
    that an older Pydantic model would accept as valid garbage.
    """

    def __init__(self, version: object) -> None:
        supported = ", ".join(str(v) for v in sorted(_SUPPORTED_SCHEMA_VERSIONS))
        super().__init__(
            f"Unsupported manifest schema version: {version!r}. "
            f"This rv release supports versions: {supported}. "
            "Upgrade rv or downgrade your manifest to a supported version."
        )


class AssetType(StrEnum):
    """Supported asset orchestration types."""

    SYMLINK = "symlink"
    COPY = "copy"
    TEMPLATE = "template"
    SECRET = "secret"


class AssetHookCommand(BaseModel):
    """Inline shell-command hook definition for a per-asset hook step."""

    command: str = Field(..., description="Shell command to execute (passed as list via shlex.split)")


class AssetHookPlugin(BaseModel):
    """Plugin-reference hook definition for a per-asset hook step."""

    plugin: str = Field(..., description="Plugin name to invoke from the repository plugins/ directory")


class AssetHooks(BaseModel):
    """Pre- and post-mutation hook definitions for a single asset."""

    pre: list[AssetHookCommand | AssetHookPlugin] = Field(
        default_factory=list,
        description="Hooks to execute before the asset mutation is applied.",
    )
    post: list[AssetHookCommand | AssetHookPlugin] = Field(
        default_factory=list,
        description="Hooks to execute after the asset mutation is applied successfully.",
    )


class ConflictStrategy(StrEnum):
    """Strategies to resolve file conflicts during restore."""

    PROMPT = "prompt"
    OVERWRITE = "overwrite"
    SKIP = "skip"
    ABORT = "abort"


class Asset(BaseModel):
    """Asset definition representing a symlink, copied file, or template."""

    id: str = Field(..., description="Unique identifier for the asset")
    type: AssetType = Field(AssetType.SYMLINK, description="Orchestration type")
    source: str = Field(..., description="Source path relative to the manifest.yaml")
    target: str | list[str] = Field(..., description="Target system path. Supports ${VAR} env interpolation.")
    permissions: str | None = Field(None, description="Octal permission string (e.g., '0644')")
    owner: str | None = Field(None, description="Owner username (null defaults to current user)")
    conflict_strategy: ConflictStrategy = Field(ConflictStrategy.PROMPT, description="Conflict resolution strategy")
    encrypted: bool = Field(False, description="Whether the asset source is encrypted (always true for secret type)")
    template_vars: dict[str, Any] | None = Field(None, description="Key-value mapping for template interpolation")
    hooks: AssetHooks = Field(default_factory=AssetHooks, description="Per-asset pre/post mutation hooks")

    @model_validator(mode="after")
    def validate_encrypted_secret(self) -> "Asset":
        """Ensure secret type has encrypted=True and secret type properties."""
        if self.type == AssetType.SECRET:
            self.encrypted = True
        return self

    @field_validator("source")
    @classmethod
    def validate_source_path(cls, v: str) -> str:
        """Prevent path traversal in source path."""
        import os

        normalized = os.path.normpath(v)
        if normalized.startswith("..") or os.path.isabs(normalized):
            raise ValueError(
                f"Source path '{v}' must be relative to the repository and not contain path traversal ('..')"
            )
        return v

    @field_validator("permissions")
    @classmethod
    def validate_permissions(cls, v: str | None) -> str | None:
        """Validate octal permission string."""
        if v is None:
            return v
        try:
            int(v, 8)
        except ValueError:
            raise ValueError(f"Permissions must be a valid octal string, got '{v}'")
        if len(v) != 4 or not v.startswith("0"):
            raise ValueError(f"Permissions must be a 4-digit octal string starting with 0, got '{v}'")
        return v


class Secret(BaseModel):
    """Secret definition decrypted via age before placement."""

    id: str = Field(..., description="Unique identifier for the secret")
    type: AssetType = Field(AssetType.SECRET, description="Orchestration type (fixed to secret)")
    source: str = Field(..., description="Encrypted source file path (typically ends in .age)")
    target: str | list[str] = Field(..., description="Target system path. Supports ${VAR} env interpolation.")
    permissions: str = Field("0600", description="Strict permissions enforced on secrets")
    owner: str | None = Field(None, description="Owner username (null defaults to current user)")
    encrypted: bool = Field(True, description="Always true for secrets")

    @model_validator(mode="after")
    def validate_secret_type(self) -> "Secret":
        """Enforce secret specific fields."""
        self.type = AssetType.SECRET
        self.encrypted = True
        # Enforce strict secret permissions if none provided
        if not self.permissions:
            self.permissions = "0600"
        return self

    @field_validator("source")
    @classmethod
    def validate_source_path(cls, v: str) -> str:
        """Prevent path traversal in source path."""
        import os

        normalized = os.path.normpath(v)
        if normalized.startswith("..") or os.path.isabs(normalized):
            raise ValueError(
                f"Secret source path '{v}' must be relative to the repository and not contain path traversal"
            )
        return v

    @field_validator("permissions")
    @classmethod
    def validate_permissions(cls, v: str) -> str:
        """Validate strict octal permissions for secrets (typically 0600 or 0700)."""
        try:
            val = int(v, 8)
        except ValueError:
            raise ValueError(f"Permissions must be a valid octal string, got '{v}'")
        if len(v) != 4 or not v.startswith("0"):
            raise ValueError(f"Permissions must be a 4-digit octal string starting with 0, got '{v}'")
        # Ensure secret permissions are secure (no world-readable/writable permissions)
        # 0700 or 0600 are standard. Let's make sure it is at least group/world restrictive.
        if (val & 0o077) != 0:
            raise ValueError(f"Secrets must have secure permissions restricting group and world access, got '{v}'")
        return v


class DockerConfig(BaseModel):
    """Docker environment provisioning."""

    images: list[str] = Field(default_factory=list)


class NodeConfig(BaseModel):
    """Node/Nvm environment provisioning."""

    version_file: str | None = Field(default=None, description="Path to .nvmrc or similar version file")
    version: str | None = Field(default=None, description="Explicit target Node.js version")


class Packages(BaseModel):
    """System and language level packages."""

    brew: list[str] = Field(default_factory=list)
    apt: list[str] = Field(default_factory=list)
    flatpak: list[str] = Field(default_factory=list)
    snap: list[str] = Field(default_factory=list)
    pacman: list[str] = Field(default_factory=list, description="Arch Linux packages via pacman")
    dnf: list[str] = Field(default_factory=list, description="Fedora/RHEL packages via dnf")
    nix: list[str] = Field(default_factory=list, description="Nix packages via nix-env (nixpkgs.<pkg>)")
    cargo: list[str] = Field(default_factory=list, description="Rust tools via cargo install")
    pip: list[str] = Field(default_factory=list, description="Python tools via pip install --user")
    docker: DockerConfig = Field(default_factory=lambda: DockerConfig())
    node: NodeConfig = Field(default_factory=lambda: NodeConfig())


class Profile(BaseModel):
    """Profile configuration linking assets, secrets, and packages."""

    extends: list[str] = Field(default_factory=list, description="Base profiles extended by this profile")
    assets: list[str | Asset] = Field(default_factory=list, description="Assets to restore (by ID or inline)")
    secrets: list[str | Secret] = Field(default_factory=list, description="Secrets to restore (by ID or inline)")
    packages: list[str] = Field(default_factory=list, description="Top-level package groups referenced by this profile")


class BackupRetentionConfig(BaseModel):
    """Controls automatic cleanup of old transaction backup snapshots."""

    max_count: int = Field(default=10, ge=1, description="Keep at most N backup snapshots (FIFO eviction)")
    max_age_days: int = Field(default=30, ge=1, description="Delete backup snapshots older than N days")


class MachineOverridesConfig(BaseModel):
    """Machine override configuration."""

    enabled: bool = Field(default=True, description="Enable machine overrides")
    path: str = Field(
        default="machine/{hostname}.yaml", description="Path pattern for host-specific override manifests"
    )


class Manifest(BaseModel):
    """Root configuration manifest representing the complete repository state."""

    version: int = Field(2, description="Manifest schema version")
    assets: list[Asset] = Field(default_factory=list, description="Global pool of assets")
    secrets: list[Secret] = Field(default_factory=list, description="Global pool of secrets")
    packages: Packages = Field(default_factory=lambda: Packages(), description="Global package definitions")
    profiles: dict[str, Profile] = Field(default_factory=dict, description="Named deployment profiles")
    machine_overrides: MachineOverridesConfig = Field(default_factory=lambda: MachineOverridesConfig())
    backup_retention: BackupRetentionConfig = Field(
        default_factory=lambda: BackupRetentionConfig(),
        description="Automatic backup snapshot pruning configuration.",
    )

    @field_validator("version")
    @classmethod
    def validate_schema_version(cls, v: int) -> int:
        """Reject manifests that declare an unsupported schema version.

        This guard runs *inside* Pydantic model_validate(), which is called
        only after ManifestLoader.load() performs the raw-version pre-check.
        The double-guard ensures correctness even when Manifest is constructed
        directly in tests or other code paths.
        """
        if v not in _SUPPORTED_SCHEMA_VERSIONS:
            raise UnsupportedSchemaVersionError(v)
        return v

File: internal/manifest/manifest.go
// Package manifest owns the manifest.yaml schema, its strict loader, and its validation rules.
// The schema is part of the compatibility contract: versions 1 and 2 must load exactly as the
// Python implementation loaded them.
package manifest

import (
	"encoding/json"
	"errors"
	"fmt"
)

// Sentinel errors. Callers branch on these with errors.Is, never on message text.
var (
	// ErrValidation covers every domain rule violation found by Validate.
	ErrValidation = errors.New("manifest validation failed")
	// ErrUnsupportedSchemaVersion is returned for any version outside {1, 2}.
	ErrUnsupportedSchemaVersion = errors.New("unsupported manifest schema version")
)

// SupportedVersions is the set of schema versions this release understands.
var SupportedVersions = []int{1, 2}

// DefaultVersion applies when the manifest omits `version`.
const DefaultVersion = 2

// AssetType is the orchestration type of an asset.
type AssetType string

// Asset types.
const (
	TypeSymlink  AssetType = "symlink"
	TypeCopy     AssetType = "copy"
	TypeTemplate AssetType = "template"
	TypeSecret   AssetType = "secret"
)

// Valid reports whether t is a known asset type.
func (t AssetType) Valid() bool {
	switch t {
	case TypeSymlink, TypeCopy, TypeTemplate, TypeSecret:
		return true
	}
	return false
}

// ConflictStrategy decides what happens when a target already exists.
type ConflictStrategy string

// Conflict strategies.
const (
	ConflictPrompt    ConflictStrategy = "prompt"
	ConflictOverwrite ConflictStrategy = "overwrite"
	ConflictSkip      ConflictStrategy = "skip"
	ConflictAbort     ConflictStrategy = "abort"
)

// Valid reports whether s is a known conflict strategy.
func (s ConflictStrategy) Valid() bool {
	switch s {
	case ConflictPrompt, ConflictOverwrite, ConflictSkip, ConflictAbort:
		return true
	}
	return false
}

// StringOrSlice holds a field written as either a scalar or a list. It remembers which, because
// the lockfile round trip must write back the same shape it read.
type StringOrSlice struct {
	Values    []string
	wasScalar bool
}

// Scalar builds a StringOrSlice that marshals back as a bare string.
func Scalar(v string) StringOrSlice { return StringOrSlice{Values: []string{v}, wasScalar: true} }

// Slice builds a StringOrSlice that marshals back as a list.
func Slice(v ...string) StringOrSlice { return StringOrSlice{Values: v} }

// IsScalar reports whether the value was written as a scalar.
func (s StringOrSlice) IsScalar() bool { return s.wasScalar }

// UnmarshalJSON accepts a string or a list of strings.
func (s *StringOrSlice) UnmarshalJSON(b []byte) error {
	var one string
	if err := json.Unmarshal(b, &one); err == nil {
		*s = StringOrSlice{Values: []string{one}, wasScalar: true}
		return nil
	}
	var many []string
	if err := json.Unmarshal(b, &many); err != nil {
		return fmt.Errorf("expected a string or a list of strings, got %s", b)
	}
	*s = StringOrSlice{Values: many}
	return nil
}

// MarshalJSON writes a scalar back as a scalar and a list back as a list.
func (s StringOrSlice) MarshalJSON() ([]byte, error) {
	if s.wasScalar && len(s.Values) == 1 {
		return json.Marshal(s.Values[0])
	}
	return json.Marshal(s.Values)
}

// Hook is one entry in an asset's pre or post hook list. Exactly one field may be set.
type Hook struct {
	Command string `json:"command,omitempty"`
	Plugin  string `json:"plugin,omitempty"`
}

// Hooks are the per-asset lifecycle hooks. They are planned during planning and executed inside
// the transaction, interleaved around their own asset's mutation.
type Hooks struct {
	Pre  []Hook `json:"pre,omitempty"`
	Post []Hook `json:"post,omitempty"`
}

// Asset is one file or directory managed by rv.
type Asset struct {
	ID               string           `json:"id"`
	Type             AssetType        `json:"type,omitempty"`
	Source           string           `json:"source"`
	Target           StringOrSlice    `json:"target"`
	Permissions      *string          `json:"permissions,omitempty"`
	Owner            *string          `json:"owner,omitempty"`
	ConflictStrategy ConflictStrategy `json:"conflict_strategy,omitempty"`
	Encrypted        bool             `json:"encrypted,omitempty"`
	TemplateVars     map[string]any   `json:"template_vars,omitempty"`
	Hooks            Hooks            `json:"hooks,omitempty"`
}

// Secret is an age-encrypted asset. Its type, encryption and permission floor are forced at load
// time, so it deliberately carries fewer fields than Asset.
type Secret struct {
	ID          string        `json:"id"`
	Type        AssetType     `json:"type,omitempty"`
	Source      string        `json:"source"`
	Target      StringOrSlice `json:"target"`
	Permissions string        `json:"permissions,omitempty"`
	Owner       *string       `json:"owner,omitempty"`
	Encrypted   bool          `json:"encrypted,omitempty"`
}

// Asset renders the secret as an Asset for the engine, which treats both uniformly.
func (s Secret) Asset() Asset {
	perms := s.Permissions
	return Asset{
		ID:               s.ID,
		Type:             TypeSecret,
		Source:           s.Source,
		Target:           s.Target,
		Permissions:      &perms,
		Owner:            s.Owner,
		ConflictStrategy: ConflictOverwrite,
		Encrypted:        true,
	}
}

// DockerConfig lists images to pull.
type DockerConfig struct {
	Images []string `json:"images,omitempty"`
}

// NodeConfig selects a node version. Version wins over VersionFile.
type NodeConfig struct {
	VersionFile string `json:"version_file,omitempty"`
	Version     string `json:"version,omitempty"`
}

// Packages is the global package pool, one list per provider.
type Packages struct {
	Brew    []string     `json:"brew,omitempty"`
	Apt     []string     `json:"apt,omitempty"`
	Flatpak []string     `json:"flatpak,omitempty"`
	Snap    []string     `json:"snap,omitempty"`
	Pacman  []string     `json:"pacman,omitempty"`
	Dnf     []string     `json:"dnf,omitempty"`
	Nix     []string     `json:"nix,omitempty"`
	Cargo   []string     `json:"cargo,omitempty"`
	Pip     []string     `json:"pip,omitempty"`
	Docker  DockerConfig `json:"docker,omitempty"`
	Node    NodeConfig   `json:"node,omitempty"`
}

// ListNames are the package groups a profile may reference that map to a flat name list.
var ListNames = []string{"brew", "apt", "flatpak", "snap", "pacman", "dnf", "nix", "cargo", "pip"}

// List returns the package list for a group name, and whether that group is a flat list.
func (p *Packages) List(group string) ([]string, bool) {
	switch group {
	case "brew":
		return p.Brew, true
	case "apt":
		return p.Apt, true
	case "flatpak":
		return p.Flatpak, true
	case "snap":
		return p.Snap, true
	case "pacman":
		return p.Pacman, true
	case "dnf":
		return p.Dnf, true
	case "nix":
		return p.Nix, true
	case "cargo":
		return p.Cargo, true
	case "pip":
		return p.Pip, true
	}
	return nil, false
}

// ProfileRef is a profile's reference to an asset or secret: either an ID from the global pool
// or an inline definition.
type ProfileRef[T any] struct {
	ID     string
	Inline *T
}

// UnmarshalJSON accepts a bare ID string or an inline object.
func (r *ProfileRef[T]) UnmarshalJSON(b []byte) error {
	var id string
	if err := json.Unmarshal(b, &id); err == nil {
		*r = ProfileRef[T]{ID: id}
		return nil
	}
	var inline T
	if err := json.Unmarshal(b, &inline); err != nil {
		return fmt.Errorf("expected an ID string or an inline definition: %w", err)
	}
	*r = ProfileRef[T]{Inline: &inline}
	return nil
}

// MarshalJSON writes a reference back in the shape it was read.
func (r ProfileRef[T]) MarshalJSON() ([]byte, error) {
	if r.Inline != nil {
		return json.Marshal(r.Inline)
	}
	return json.Marshal(r.ID)
}

// Profile names a subset of assets, secrets and package groups, optionally extending others.
type Profile struct {
	Extends  []string             `json:"extends,omitempty"`
	Assets   []ProfileRef[Asset]  `json:"assets,omitempty"`
	Secrets  []ProfileRef[Secret] `json:"secrets,omitempty"`
	Packages []string             `json:"packages,omitempty"`
}

// BackupRetention bounds how many transaction snapshots survive, and for how long.
type BackupRetention struct {
	MaxCount   *int `json:"max_count,omitempty"`
	MaxAgeDays *int `json:"max_age_days,omitempty"`
}

// MachineOverrides configures per-host manifest overrides.
type MachineOverrides struct {
	Enabled *bool  `json:"enabled,omitempty"`
	Path    string `json:"path,omitempty"`
}

// Manifest is the root declaration at the workspace root.
type Manifest struct {
	Version          *int               `json:"version,omitempty"`
	Assets           []Asset            `json:"assets,omitempty"`
	Secrets          []Secret           `json:"secrets,omitempty"`
	Packages         Packages           `json:"packages,omitempty"`
	Profiles         map[string]Profile `json:"profiles,omitempty"`
	MachineOverrides MachineOverrides   `json:"machine_overrides,omitempty"`
	BackupRetention  BackupRetention    `json:"backup_retention,omitempty"`
}

// SchemaVersion returns the declared version, or DefaultVersion when it was omitted.
func (m *Manifest) SchemaVersion() int {
	if m.Version == nil {
		return DefaultVersion
	}
	return *m.Version
}

// OverridesEnabled reports whether machine overrides apply. They default to on.
func (m *Manifest) OverridesEnabled() bool {
	return m.MachineOverrides.Enabled == nil || *m.MachineOverrides.Enabled
}

// Retention returns the effective retention bounds, applying the documented defaults.
func (m *Manifest) Retention() (maxCount, maxAgeDays int) {
	maxCount, maxAgeDays = 10, 30
	if m.BackupRetention.MaxCount != nil {
		maxCount = *m.BackupRetention.MaxCount
	}
	if m.BackupRetention.MaxAgeDays != nil {
		maxAgeDays = *m.BackupRetention.MaxAgeDays
	}
	return maxCount, maxAgeDays
}
